flocker/control/_persistence.py

```diff
@@ -30,10 +30,110 @@
 
+# The latest configuration version. Configuration versions are
+# always integers.
+_CONFIG_VERSION = 2
+
+# Map of serializable class names to classes
+_CONFIG_CLASS_MAP = {cls.__name__: cls for cls in SERIALIZABLE_CLASSES}
+
+
+class ConfigurationMigrationError(Exception):
+    """
+    Error raised when a configuration migration is unable to
+    complete successfully.
+    """
+
+
+class MissingMigrationError(Exception):
+    """
+    Error raised when a configuration migration method cannot be found.
+    """
+    def __init__(self, source_version, target_version):
+        """
+        Initialize a missing migration exception.
+
+        :param int source_version: The version to migrate from.
+        :param int target_version: The version to migrate to.
+        """
+        self.source_version = source_version
+        self.target_version = target_version
+        self.message = (
+            u"Unable to find a migration path for a version {source} "
+            u"to version {target} configuration. No migration method "
+            u"upgrade_from_v{source} could be found.".format(
+                source=self.source_version, target=self.target_version
+            )
+        )
+        super(MissingMigrationError, self).__init__(self.message)
+
+
+def migrate_configuration(source_version, target_version,
+                          config, migration_class):
+    """
+    Migrate a persisted configuration from one version to another
+    in sequential upgrades, e.g. a source version of 1 and target
+    version of 3 will perform two upgrades, from version 1 to 2,
+    followed by 2 to 3.
+
+    Calls the correct ``migration_class`` class methods for
+    sequential upgrades between the suppled source and target versions.
+
+    :param int source_version: The version to migrate from.
+    :param int target_version: The version to migrate to.
+    :param bytes config: The source configuration blob.
+    :param class migration_class: The class containing the methods
+        that will be used for migration.
+
+    :return bytes: The updated configuration blob after migration.
+    :raises MissingMigrationError: Raises this exception if any of the
+        required upgrade methods cannot be found in the supplied migration
+        class, before attempting to execute any upgrade paths.
+    """
+    upgraded_config = config
+    current_version = source_version
+    migrations_sequence = []
+    for upgrade_version in range(source_version + 1, target_version + 1):
+        with _LOG_UPGRADE(configuration=upgraded_config,
+                          source_version=current_version,
+                          target_version=upgrade_version):
+            migration_method = u"upgrade_from_v%d" % current_version
+            try:
+                migration = getattr(migration_class, migration_method)
+                migrations_sequence.append(migration)
+                current_version = current_version + 1
+            except AttributeError:
```

Comment:
As a general rule, you want to make `try` clauses as small as possible in order to reduce the likelihood of accidentally catching something that you didn't mean to, because that can lead to masking serious errors and making code much harder to debug.

This is **especially** true for `AttributeError`, which is often raised due to programming mistakes (e.g. `self.dataRecieved(some_data)`).

The right thing to do here is either:
  1. Restrict the `try` block to just the line with the `getattr`
  2. Don't use `try` at all. Instead `getattr(migration_class, migration_method, None)` and then raise the exception if `migration` is `None`.